src/audio/pipeline.c

```diff
@@ -972,10 +974,13 @@ static uint64_t pipeline_task(void *arg)
 			trace_pipe_error_with_ids(p, "pipeline_task(): xrun "
 						  "recover failed! pipeline "
 						  "will be stopped!");
-			return 0; /* failed - host will stop this pipeline */
+			/* failed - host will stop this pipeline */
+			return pipeline_is_timer_driven(p) ?  0 :
+			SOF_TASK_STATE_COMPLETED;
 		}
 	}
 
 	tracev_pipe("pipeline_task() sched");
-	return p->ipc_pipe.period;
+	return  pipeline_is_timer_driven(p) ? p->ipc_pipe.period :
```

Comment:
@mrajwa What do you think about:
`return  pipeline_is_timer_driven(p) ? p->ipc_pipe.period : 0`
and then in edf_schedule:
`if (ret == 0) {`

Looks simpler and we won't need to modify xrun code above (can still return 0).

Reply:
Okay, let's code it this way.